static-analysis/teammates-mackerTest.xml

```diff
@@ -0,0 +1,159 @@
+                </allow>
+                
+                <!--Violations that need fixing-->
+                <allow>
+                    <from class="teammates.test.cases.logic.StudentsLogicTest" />
+                    <to class="teammates.storage.entity.CourseStudent" />
+                </allow>
+                <allow>
+                    <from class="teammates.test.cases.storage.FeedbackResponseCommentsDbTest" />
+                    <to class="teammates.storage.entity.FeedbackResponseComment" />
+                </allow>
+            </deny>
+        </access-rule>
+    </ruleset>
+    <ruleset name="TestDriverComponent: Only certain test cases can access GaeSimulation/BackDoorLogic">
+        <access-rule>
+            <deny>
+                <to>
+                    <include class="teammates.test.driver.GaeSimulation" />
+                    <include class="teammates.logic.backdoor.BackDoorLogic" />
+                </to>
+                <allow>
+                    <from>
+                        <include class="teammates.test.cases.storage.**" />
+                        <include class="teammates.test.cases.logic.**" />
+                        <include class="teammates.test.cases.action.**" />
+                        <include class="teammates.test.cases.automated.**" />
+                        <include class="teammates.test.cases.search.**" />
+                        <include class="teammates.test.cases.BaseComponentTestCase" />
+                        
+                        <!--Violation that needs fixing-->
+                        <include class="teammates.test.cases.pagedata.StudentFeedbackResultsPageDataTest" />
+                    </from>
+                </allow>
+                <allow>
+                    <!-- to import sample data for new instructor -->
+                    <from class="teammates.ui.controller.AdminInstructorAccountAddAction" />
+                    <to class="teammates.logic.backdoor.BackDoorLogic" />
+                </allow>
+                <allow>
+                    <from class="teammates.logic.backdoor.BackDoorServlet" />
+                    <to class="teammates.logic.backdoor.BackDoorLogic" />
+                </allow>
+            </deny>
+        </access-rule>
+    </ruleset>
+    <ruleset name="TestDriverComponent: BackDoor test driver is only for browser test">
+        <access-rule>
+            <deny>
+                <to class="teammates.test.driver.BackDoor" />
+                <allow>
+                    <from>
+                        <include class="teammates.test.cases.browsertests.**" />
+                        <include class="teammates.test.driver.BackDoor**$**" />
+                        <include class="teammates.test.cases.testdriver.BackDoorTest" />
+                        
+                        <!--Violation that needs fixing-->
+                        <include class="teammates.test.cases.BaseTestCaseWithDatastoreAccess" />
```

Comment:
I think you're thinking of something else, namely some way such that `BaseTestCaseWithDatastoreAccess` would not have dependency to `BackDoor`, right?

The design that `BaseComponentTestCase` extends `BaseTestCaseWithDatastoreAccess` is done precisely for the reason you provided. If you want to improve it such that the superclass does not have access to `BackDoor`, the way is:

```
BaseTestCaseWithDatastoreAccess <- Base*TestCase
               ^
               |
     BaseComponentTestCase
```
where `Base*TestCase` is some class which I can't really think of a good name for now. That means, all the methods `get*` in `BaseTestCaseWithDatastoreAccess` will be abstract, which still justifies all the `@Override` in `BaseComponentTestCase`.